### PR TITLE
Add iota->broadcast case to iota->sort->slice mapping for TopK

### PR DESCRIPTION
**Pull request by NatashaKnk**

### CODE REVIEW

File: compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp [hunk 1]
```diff
@@ -1540,6 +1540,40 @@ struct CustomCallIsTopK final
...
+// broadcasts where the last dimension of the iota is preserved throughout.
+bool isIotaOrIotaBroadcast(PatternRewriter &rewriter, Value input) {
+  auto iotaOp =
+      dyn_cast_or_null<mlir::stablehlo::IotaOp>(input.getDefiningOp());
+  if (iotaOp) {
+    int64_t iotaDim = iotaOp.getIotaDimension();
+    auto iotaLastDim = cast<ShapedType>(iotaOp.getType()).getRank() - 1;
+    if (iotaDim == iotaLastDim) {
+      return true;
+    } else {
```
Comment (rsuderman) [inline compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp]: You can get rid of the `else` part as we returned in the `if` case.

File: compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp [hunk 2]
```diff
@@ -1540,6 +1540,40 @@ struct CustomCallIsTopK final
   }
 };
 
+// Recursive helper function that identifies an Iota followed by a set of
+// broadcasts where the last dimension of the iota is preserved throughout.
+bool isIotaOrIotaBroadcast(PatternRewriter &rewriter, Value input) {
+  auto iotaOp =
+      dyn_cast_or_null<mlir::stablehlo::IotaOp>(input.getDefiningOp());
+  if (iotaOp) {
```
Comment (rsuderman) [inline compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp]: I usually do the 
```
if (auto iotaOp = dyn_cast_or_null<mlir::stablehlo::IotaOp>(input.getDefiningOp())) {
...
}
```

File: compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp [hunk 3]
```diff
@@ -1540,6 +1540,40 @@ struct CustomCallIsTopK final
...
+    } else {
+      (void)rewriter.notifyMatchFailure(iotaOp,
+                                        "Iota must be on last dimension");
+      return false;
+    }
+  }
+
+  auto broadcastOp = dyn_cast_or_null<mlir::stablehlo::BroadcastInDimOp>(
+      input.getDefiningOp());
+  if (broadcastOp) {
```
Comment (rsuderman) [inline compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp]: Ditto

File: compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp (hunk 1 above)
Comment (NatashaKnk) [reply compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp]: Good point, thanks!

**Review by rsuderman — APPROVED** [review]

**Pull request merged**